File: Programmers/SeungDo/3주차(Stack, Queue)/기능개발.cpp
#include <vector>
// 큐에 있는 모든 작업 진도에 작업 속도를 더한 후 반복문을 이용해 맨 앞에 있는 작업이 100이 넘었는지 확인 후 넘었다면 팝 하는 방식으로 풀이 해봤습니다.
using namespace std;

vector<int> solution(vector<int> progresses, vector<int> speeds) {
    vector<int> answer;
    /* 
       변수 x를 둔 이유는 for문의 조건을 progresses.size()로 뒀을 때 팝이 되면 
       큐의 크기가 for문을 돌릴 때마다 -1이 되기 때문에 for문이 작업을 다 비교하지 못하고 끝나버림
       그래서 큐의 처음 크기를 나타내는 변수 x를 두어 끝난 작업을 배포한 뒤에 
       count만큼 x값을 빼는 방식으로 큐 안에 있는 모든 작업을 다 비교할수 있게 함   
    */   
    int x = progresses.size(); // 작업 진도가 들어있는 큐의 크기 저장 
    while (1) {
        int count = 0; // 몇 개가 배포됐는지 count
        for (int i = 0; i < x; i++) progresses[i] += speeds[i]; //모든 작업 진도에 작업 속도를 더함
        for (int i = 0; i < x; i++) {
            if (progresses.front() >= 100) { // 맨앞에 있는 작업의 진도가 100을 넘었다면 
                count++; // 배포할 작업 수 ++
                progresses.erase(progresses.begin()); // 맨앞 작업 진도 팝
                speeds.erase(speeds.begin()); // 맨앞 작업 속도 팝
            }
            else break;
        }// 맨앞에 있는 작업의 진도가 100을 넘지않았을때는 바로 break  
        if (count > 0) { // 배포할 작업 수가 1개 이상이라면 
            answer.push_back(count); // answer에 count값 푸쉬
            x -= count; // 배포한 수 만큼 전체크기를 줄임
        }
        if (progresses.empty() == 1) return answer; // 큐가 비었다면 answer 리턴
    }
}

File: 기능개발.java
import java.util.ArrayList;

class Solution {
    public int[] solution(int[] progresses, int[] speeds) {
        int[] answer = {};
        int index = 0; //아직 배포되지 않은 작업의 가장 앞 작업을 가리키는 index
        int count = 0;
        ArrayList<Integer> arr = new ArrayList<>();

        while (index != progresses.length && progresses[index] < 100) { //index != progresses.length를 안 쓰면 index of bounds exception
            for (int i = 0; i < progresses.length; i++) //각 작업에 speeds 만큼식 더해준다.
                progresses[i] += speeds[i];

            while (index != progresses.length && progresses[index] >= 100) { //더해준 후 index 부분이 100 이상이면
                count++; //count를 올려주고
                index++; //다음 인덱스 가리킨다.
                if (index != progresses.length && progresses[index] < 100) { //가리키는 index의 작업이 100보다 작으면
                    arr.add(count); //count 만큼을 arr에 추가한다.
                    count=0; //count 0으로 다시 초기화
                }
                //Index out of bounds exception를 방지하기 위해 마지막 인덱스 였었던 경우를 따로 나눈다.
                else if (index==progresses.length&&progresses[index-1]>=100) arr.add(count);
            }
        }

        answer=new int[arr.size()];
        for (int i = 0; i < arr.size(); i++)
            answer[i]=arr.get(i);


        return answer;
    }
}